validphys2/src/validphys/covmats.py

```diff
@@ -339,6 +342,102 @@ def dataset_inputs_t0_covmat_from_systematics(
+            generic_path = pathlib.Path("datacuts_theory_theorycovmatconfig_user_covmat.csv")
+        else:
+            generic_path = pathlib.Path("datacuts_theory_theorycovmatconfig_theory_covmat_custom.csv")
+    theorypath = output_path/"tables"/generic_path
+    theory_covmat = pd.read_csv(theorypath, index_col=[0, 1, 2], header=[0, 1, 2], sep="\t|,", engine="python").fillna(0)
+    return theory_covmat.values
+    
+
+def dataset_inputs_total_covmat(dataset_inputs_loaded_cd_with_cuts,
+    *,
+    data_input,
+    use_weights_in_covmat=True,
+    norm_threshold=None,
+    dataset_inputs_t0_predictions,
+    loaded_theory_covmat,
+    theory_covmat_flag,
+    use_thcovmat_in_sampling,
+    use_t0_sampling
+    ):
+    if use_t0_sampling is False:
+        covmat = dataset_inputs_t0_total_covmat(
+        dataset_inputs_loaded_cd_with_cuts,
+        data_input=data_input,
+        use_weights_in_covmat=use_weights_in_covmat,
+        norm_threshold=norm_threshold,
+        dataset_inputs_t0_predictions=None,
+        loaded_theory_covmat=loaded_theory_covmat,
+        theory_covmat_flag=theory_covmat_flag,
+        use_thcovmat_in_fitting=use_thcovmat_in_sampling,
+        use_t0_fitting=use_t0_sampling
+    )
+    else:
+        covmat = dataset_inputs_t0_total_covmat(
+        dataset_inputs_loaded_cd_with_cuts,
+        data_input=data_input,
+        use_weights_in_covmat=use_weights_in_covmat,
+        norm_threshold=norm_threshold,
+        dataset_inputs_t0_predictions=dataset_inputs_t0_predictions,
+        loaded_theory_covmat=loaded_theory_covmat,
+        theory_covmat_flag=theory_covmat_flag,
+        use_thcovmat_in_fitting=use_thcovmat_in_sampling,
+        use_t0_fitting=use_t0_sampling
+    )
+    
+    return covmat
+
+
+def dataset_inputs_t0_total_covmat(dataset_inputs_loaded_cd_with_cuts,
+    *,
+    data_input,
+    use_weights_in_covmat=True,
+    norm_threshold=None,
+    dataset_inputs_t0_predictions,
+    loaded_theory_covmat,
+    theory_covmat_flag,
+    use_thcovmat_in_fitting,
+    use_t0_fitting
+    ):
+    if use_t0_fitting is False:
+        exp_covmat = dataset_inputs_covmat_from_systematics(
+        dataset_inputs_loaded_cd_with_cuts,
+        data_input,
+        use_weights_in_covmat,
+        norm_threshold=norm_threshold,
+        _list_of_central_values=None
+    )
+    else:
+        exp_covmat = dataset_inputs_covmat_from_systematics(
+        dataset_inputs_loaded_cd_with_cuts,
+        data_input,
+        use_weights_in_covmat,
+        norm_threshold=norm_threshold,
+        _list_of_central_values=dataset_inputs_t0_predictions
+    )
+    if theory_covmat_flag and use_thcovmat_in_fitting:
+            #Adding regularization to theory covmat to make sqrt possible  
+            diag_enha = 1.e-6
+            return exp_covmat + loaded_theory_covmat*(np.ones_like(loaded_theory_covmat) + diag_enha*np.eye(loaded_theory_covmat.shape[0]))
+    return exp_covmat
```

Comment:
I think for these two functions (and all the branches it created) we need to rethink a bit the names and what we want to have.

Imho we should have `sampling_total_covmat` and `fitting_total_covmat` (or something like that) so that it's clear what they should be used for.

Personally I would just enforce that `sampling` does not include `t0` and `fitting` does but yesterday talking with @Radonirinaunimi he was interested in having the possibility of setting that flag on and off so we might as well included (since you already did).

We can sketch the different names / branches that we want in a blackboard when I'm at the office, it will be faster.

Reply:
Ok sure. Till then I leave the names as they are now

Reply:
BTW I believe that we should leave the flags for t0 as well